x-pack/plugins/event_log/server/plugin.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { first } from 'rxjs/operators';
+import {
+  CoreSetup,
+  CoreStart,
+  Logger,
+  Plugin as CorePlugin,
+  PluginInitializerContext,
+  ClusterClient,
+} from 'src/core/server';
+
+import { IEventLogConfig, IEventLogService, IEventLogger, IEventLogConfig$ } from './types';
+import { EventLogService } from './event_log_service';
+import { createEsContext, EsContext } from './es';
+
+export type PluginClusterClient = Pick<ClusterClient, 'callAsInternalUser' | 'asScoped'>;
+
+// TODO - figure out how to get ${kibana.index} for `.kibana`
+const KIBANA_INDEX = '.kibana';
+
+const PROVIDER = 'event_log';
+const ACTIONS = {
+  starting: 'starting',
+  stopping: 'stopping',
+};
+
+export class Plugin implements CorePlugin<IEventLogService> {
+  private readonly config$: IEventLogConfig$;
+  private systemLogger: Logger;
+  private eventLogService?: IEventLogService;
+  private esContext?: EsContext;
+  private eventLogger?: IEventLogger;
+
+  constructor(private readonly context: PluginInitializerContext) {
+    this.systemLogger = this.context.logger.get();
+    this.config$ = this.context.config.create<IEventLogConfig>();
+  }
+
+  async setup(core: CoreSetup): Promise<IEventLogService> {
+    this.systemLogger.debug('setting up plugin');
+
+    const config = await this.config$.pipe(first()).toPromise();
+
+    this.esContext = createEsContext({
+      logger: this.systemLogger,
+      // TODO: get index prefix from config.get(kibana.index)
+      indexNameRoot: KIBANA_INDEX,
+      clusterClient$: core.elasticsearch.adminClient$,
+    });
+
+    this.eventLogService = new EventLogService({
+      config,
+      esContext: this.esContext,
+      systemLogger: this.systemLogger,
+    });
+
+    this.eventLogService.registerProviderActions(PROVIDER, Object.values(ACTIONS));
+
+    this.eventLogger = this.eventLogService.getLogger({
+      event: { provider: PROVIDER },
+    });
+
+    return this.eventLogService;
+  }
+
+  async start(core: CoreStart) {
+    this.systemLogger.debug('starting plugin');
+
+    // launches initialization async
+    if (this.eventLogService!.isIndexingEntries()) {
```

Comment:
Is there a way we can avoid the `!`?
We won't know if a logical error causes us to call `start` before `setup` until run time... just wondering if we can restructure things to make it impossible to accidentally do that?

Reply:
I'd like to find a way to do this.  It's tough since it's an instance variable which by definition cannot be initialized in the constructor.  I think there's probably a pattern of creating yet another object for some indirection so we don't have to deal with this, but then you're adding code just to make the type-y safe-y bits more pure, which seems not right.  I'll also check the existing NP plugins again as well.

Reply:
Yeah, I've encountered the same issue with moving TM to NP....
Seems a general issue with using Classes as the base for things that can't be provided in the constructor.

Quite annoying and means we're very likely to encounter null pointer exceptions... worth considering a adopting Optionals throughout this code to avoid these nulls 😬

Reply:
I changed these so there is an explicit check for the references not having been set, TS is then smart enough to know you don't need the `!` afterwards.

eg, avoids having to use `this.esContext!.initialize()`

```js
    if (!this.esContext) throw new Error('esContext not initialized');
    ...
    this.esContext.initialize()
```